_posts/2022-09-08-patterns-with-suspense.markdown

```diff
@@ -0,0 +1,589 @@
+
+And that's it. Looking at our `Artist` component again we can see how it all
+fits together:
+
+```tsx
+const artistData = prepareArtistData()
+
+export const Artist = () => {
+  const data = artistData.read() // throw promise, throw error or return result
+
+  return (
+    <>
+      {data && (
+        <>
+          {data.name}, {data.birthday}
+        </>
+      )}
+
+      <Artworks />
+    </>
+  )
+}
+```
+
+No `awaiting` anything, and nothing to slow your app code execution down
+because, again, data and code have been effectively separated. On `render` the
+whole component tree executes and areas that need loading states simply
+_display_ the loading state via `Suspense` boundary, not block execution. And if
+you need formal error handling, wrap `<Suspense>` in an additional [Error
+Boundary][error] for display.
+
+Here's another pseudocode example, using a router which preloads all route data
+ahead of time:
+
+```tsx
+const routes = [
+  {
+    path: '/':
+    queryRef: loadQuery(homeQuery)
+    Component: () => {
+      const data = usePreloadedQuery(homeQuery)
+      ...
+    }
+  },
+  {
+    path: '/artists':
+    queryRef: loadQuery(artistsQuery)
+    Component: () => {
+      const data = usePreloadedQuery(artistsQuery)
+      ...
+    }
+  }
+]
+
+bootApp({ routes })
+```
+
+This pattern is _flexible_. One can imagine any number of ways we could
+`prepare()` this data to be `read()`. And underneath it all the mechanism is
+_simple_; its just us throwing a promise to suspend, throwing an error, or
+returning data to be used in the component. Very cool.
+
+## Gotchas and Emerging Patterns Around Them
+
+Now that suspense is properly understood lets get to some common patterns that
+one can use to get around the gotchas. Suspense is new; it's like the transition
+period when Functional Components (FCs) came out and people were moving from
+class based components to hooks. And remember when render props were a thing?
+They were a way to use dependency injection to "inject" props into a component,
+allowing us to share data dependencies across a react tree. That pattern was
+cool at the time but it wasn't without its headaches, and the community hadn't
+yet settled on contexts as they're now commonly understood. It was a pattern
+that we had to come up with. It's the same thing with Suspense. React has
+provided the tools for us to work with concurrency patterns in our UI, but since
+Suspense is so new we still need to figure out just how to do it right.
+
+The following bits of code have been made generic for demonstration purposes,
+but if you use [Relay][relay] (like we do at Artsy) you're officially in a
+suspense-based world and all of the following patterns will apply.
+
+### An example
+
+Say you've got an `<Autocomplete />` component that allows you to enter some
+text which fetches data to render a dropdown list. The data is being requested
+via a hook that uses suspense, and the request is being triggered on each
```

Comment:
> The data is being requested via a hook that uses suspense

**thought:** I am wondering here if the hook is using Suspense or if it's the opposite here. I was under the impression that Suspense is aware of the hook that fetches the data

Reply:
Technically speaking react is unaware of the hook until it throws a promise, at which point the promise is registered by react suspense. And the hook isn't doing anything _other_ than throwing a promise, which is just a js pattern. Its neatly decoupled

Reply:
Thanks for explaining Chris 🙏 I wasn't getting it right

Reply:
Ah, good suggestion, will do 👍